### PR TITLE
Allowing inferred searching using facets as keywords

### PR DESCRIPTION
**Pull request by bistline**
Non-numeric faceted searches will now attempt to "infer" results by converting filter values into exact phrases (using filter display names, e.g. "tuberculosis", or "Homo sapiens") and running a secondary search after all exact matches are found.  These studies are labeled with `inferred_match: true` in the response, and ordered after all exact matches have returned.  They will also list what terms matched using the standard `term_matches` parameter.  These results are weighted together, so higher confidence inferred matches are returned first (but still after all exact matches).  This PR also DRYs up the search logic somewhat by extracting MongoDB-based searches into a static method.

This PR satisfies SCP-2205.

### CODE REVIEW

File: app/controllers/api/v1/search_controller.rb [hunk 1]
```diff
@@ -557,6 +574,27 @@ def self.escape_terms_for_regex(term_list:)
         /(#{escaped_terms.join('|')})/i
       end
 
+      # generate a Mongoid::Criteria object to perform a keyword/exact phrase search based on use case
+      # supports the following cases: :keyword (individual terms), :phrase (quoted phrases & keywords)
+      # and :inferred (converting a facet-based query to keywords)
+      # will scope the query based off of :base_studies, and include/exclude studies matching
+      # :accessions based on the :query_case (included by default, but excluded in :inferred to avoid duplicates)
+      def self.generate_mongo_query_by_case(terms:, base_studies:, accessions:, query_case:)
```
Comment (devonbush) [inline app/controllers/api/v1/search_controller.rb]: hmmm.. I'm trying to think of a better word to use than 'case', since I at first thought this was referring to whether the search was case-sensitive. 'type'? 'term_type'? 'goal'? I'm not sure any of those are better though

File: app/controllers/api/v1/search_controller.rb [hunk 2]
```diff
@@ -612,6 +650,19 @@ def self.generate_bq_query_string(facets)
         with_statement + base_query + from_clause + " WHERE " + where_clauses.join(" AND ")
       end
 
+      # convert a list of facet filters into a keyword search for inferred matching
+      def self.convert_filters_for_inferred_search(facets:)
+        filter_terms = []
+        facets.each do |facet|
+          search_facet = SearchFacet.find(facet[:object_id])
+          # only use non-numeric facets
```
Comment (devonbush) [inline app/controllers/api/v1/search_controller.rb]: nice catch :). That's definitely an edge case I would have missed

File: app/controllers/api/v1/search_controller.rb [hunk 3]
```diff
@@ -612,6 +650,19 @@ def self.generate_bq_query_string(facets)
...
+          if !search_facet.is_numeric?
+            filter_terms += search_facet[:filters].map {|filter| filter[:name]}
```
Comment (devonbush) [inline app/controllers/api/v1/search_controller.rb]: I think this can be rewritten as 
```suggestion
            filter_terms += search_facet.filters.map(&:name)
```

File: test/api/search_controller_test.rb [hunk 4]
```diff
@@ -245,4 +245,38 @@ class SearchControllerTest < ActionDispatch::IntegrationTest
...
+                 "Did not find expected accessions after inferred search, expected #{inferred_accessions} but found #{json['matching_accessions']}"
+    inferred_study = json['studies'].last # inferred matches should be at the end
+    assert inferred_study['inferred_match'],
+           "Did not mark last search results as inferred_match: #{inferred_study['inferred_match']} != true"
+
+    # reset description so other tests aren't broken
+    other_study.update(description: original_description)
+
+    puts "#{File.basename(__FILE__)}: #{self.method_name} successful!"
+  end
```
Comment (devonbush) [inline test/api/search_controller_test.rb]: This is good.  Can you also add a test for 
1. Searching for multiple facets/filters behaves as expected
2. Searching for an exact phrase + facet behaves as expected.  
I expect both of these cases to work, but given that this is home-page behavior, an extra layer of testing would be good

File: app/controllers/api/v1/search_controller.rb (hunk 3 above)
Comment (bistline) [reply app/controllers/api/v1/search_controller.rb]: Actually it cannot.  The shorthand block notation does not work with hash keys (I wish it did).

File: test/api/search_controller_test.rb (hunk 4 above)
Comment (bistline) [reply test/api/search_controller_test.rb]: Can do.

Comment (bistline) [reply test/api/search_controller_test.rb]: To be clear, we already have a test for the first case.  The existing test `'should return search results using facets'` uses multiple facets at the same time.  But I can add a test for the second case.

File: app/controllers/api/v1/search_controller.rb (hunk 1 above)
Comment (bistline) [reply app/controllers/api/v1/search_controller.rb]: I landed on `generate_mongo_query_by_context`, and changed the last parameter to `query_context`.

File: test/api/search_controller_test.rb (hunk 4 above)
Comment (bistline) [reply test/api/search_controller_test.rb]: 70f04a9

File: app/controllers/api/v1/search_controller.rb (hunk 1 above)
Comment (devonbush) [reply app/controllers/api/v1/search_controller.rb]: nice

**Comment by bistline** [comment]
> I don't know what our conventions/log setups are, but this feels more like a trace/debug log than INFO

I hear you, but to be perfectly honest the logic behind searches now is somewhat confusing.  The presence of keyword + facets + inferred search led to some corner cases that I didn't expect.  So those statements started out as debug, but then I decided they will be useful for addressing issues in production where we think we should be getting results but might not.

**Comment by devonbush** [comment]
> I don't see the upside to that, really. We're trying to test the corner cases of inferred search, so I don't see the difference between extra assertions in a test that is already targeting that feature and extracting it out to its own test. I'd have to do the same setup that I did in this test.

the benefit is that when something breaks, you get a full picture of what's broken and still working, rather than only learning what the first failure was.  Concur that it's non-essential in this case

**Comment by bistline** [comment]
> > I don't see the upside to that, really. We're trying to test the corner cases of inferred search, so I don't see the difference between extra assertions in a test that is already targeting that feature and extracting it out to its own test. I'd have to do the same setup that I did in this test.
> 
> the benefit is that when something breaks, you get a full picture of what's broken and still working, rather than only learning what the first failure was. Concur that it's non-essential in this case

Fair enough - I'll abstract it out to a separate test
